### ISSUE TITLE
PopUp is not opening with activated URL Filter

### ISSUE THREAD

**Issue by kollbrt** [SafeExamBrowser/seb-mac#324]
Hi,
We have still problems with SEB and Mac when we activate the URL-Filter in OpenOlat. We can block or allow URL, this is no problem, but when students have to open a popup of a WordFile, the file doesn't open. We already checked, which URL Mac is using in the background to open the popup, but it doesn't work.
Do you have any clue, why this function is not possible? I remember, that we had the same problem some time ago also with windows and SEB, but then you could fix it. So my question: can you also fix this one? 
I attached the list with URL we were allowing (with this list the URL Filter and popup window with windows laptops worked) and the screenshots with the problem during an exam with mac.

Thank you for helping us.

Kind regards
Thomas Kollbrunner, teacher in science, BBW
![20231006_154834](https://github.com/SafeExamBrowser/seb-mac/assets/74318647/b3ab6a99-9800-438f-93a6-3372650f699e)
![20231006_154939](https://github.com/SafeExamBrowser/seb-mac/assets/74318647/8898552d-7d10-46d3-85c6-d31252b44acc)
[Liste_Olat_Word_DPrf.docx](https://github.com/SafeExamBrowser/seb-mac/files/12831852/Liste_Olat_Word_DPrf.docx)

**Comment by danschlet** [comment]
Was it this issue: https://github.com/SafeExamBrowser/seb-win-refactoring/issues/240 ?

That would have been another issue with popup windows in general, not connected to URL filters. Did you try if the word file popup works in the current Mac version when not using URL filters?

**Comment by kollbrt** [comment]
Yes, we tried already without the URL Filter and then, Mac is able to open the word file popup. I think, it is a similar problem as issue240, but I'm not sure.

**Comment by kollbrt** [comment]
Hi danschlet,
I have here a log file from one of my tests with the same problem. You can find the blocked URL at 15:45:27 and 29
I tried it two times.
It shows "This clicked link was blocked by the URL filter: (null)"
I remember, that we had the same problem with windows before you fixed that issue for SEB win.
I could give you a temporal account to our Olat Platform if this helps, so please write me an Email on thomas.kollbrunner@bbw.ch, so I could answer you there.

Pinch-to-zoom doesn't work during the test, so I assume we have the classic WebView.

[org.safeexambrowser.SafeExamBrowser+2023-10-27--13-45-11-249.log](https://github.com/SafeExamBrowser/seb-mac/files/13189701/org.safeexambrowser.SafeExamBrowser%2B2023-10-27--13-45-11-249.log)

Thanks for the help.

**Comment by danschlet** [comment]
Thanks to the log file, I was able to find a bug. The fix will be contained in the final version of SEB 3.3.1, which I attempt to release in the next couple of days.

**Comment by kollbrt** [comment]
We tried last saturday with the Pre-Release of the SEB 3.3.1, but our setting still doesn't work. I have again two log-files hier, maybe they will help in some way.
[org.safeexambrowser.SafeExamBrowser 2023-11-18--09-21-24-522.log](https://github.com/SafeExamBrowser/seb-mac/files/13427371/org.safeexambrowser.SafeExamBrowser.2023-11-18--09-21-24-522.log)
[org.safeexambrowser.SafeExamBrowser 2023-11-18--09-41-27-865.log](https://github.com/SafeExamBrowser/seb-mac/files/13427372/org.safeexambrowser.SafeExamBrowser.2023-11-18--09-41-27-865.log)

**Comment by danschlet** [comment]
The fix is contained in release candidate 2 of SEB 3.3.1, which should be released in the next couple of days.

**Comment by kollbrt** [comment]
perfect, we'll try it again.

**Comment by danschlet** [comment]
Took a bit more time than expected, but the issue should be fixed in [SEB 3.3.2 for macOS](https://safeexambrowser.org/download_en.html#MacOSX). Would be great if you could test if this works for you.

**Comment by kollbrt** [comment]
We tested it again. But no success. I have again two log-files attached. If I configured SEB directly on the mac interface, it showed me, that the button to open the word file would close SEB (foto also attached). 
![20240126_135454](https://github.com/SafeExamBrowser/seb-mac/assets/74318647/cb534a3f-84aa-4623-915f-d44c128a0496)
[org.safeexambrowser.SafeExamBrowser 2024-01-26--12-57-32-247.log](https://github.com/SafeExamBrowser/seb-mac/files/14065377/org.safeexambrowser.SafeExamBrowser.2024-01-26--12-57-32-247.log)
[org.safeexambrowser.SafeExamBrowser 2024-01-26--13-00-44-325.log](https://github.com/SafeExamBrowser/seb-mac/files/14065379/org.safeexambrowser.SafeExamBrowser.2024-01-26--13-00-44-325.log)

**Comment by danschlet** [comment]
Please note that SEB doesn't support closing the main browser window (the one, where the Start URL was initially loaded in). In case OpenOlat tries this, it should be though blocked (and not lead to the quit confirmation to be displayed). I will investigate further.

**Comment by danschlet** [comment]
According to the first log file, when you click on the link "CHEMIE BMS KO - Copy of Aufgabe", the [Quit Link](https://safeexambrowser.org/macosx/mac_usermanual_en.html#ExamPane) is invoked (Log: "Quit Link invoked"). So it seems that instead of loading this popup of a WordFile, the URL is invoked which in your settings is set for the Quit Link. This sounds improbably, maybe someone set a non-URL as Quit Link (the Quit Link **must be a properly formatted URL**, although it doesn't have to exist, doesn't have to point to an existing web page).

To analyze the setting further, I would need either or maybe both:

- Used SEB settings. Make sure to either send the configuration file with all passwords removed or indicate the used settings/exam, SEB admin and SEB quit passwords.
- It might be necessary to reproduce the issue in OpenOlat. Unfortunately I can't debug SEB for macOS if [the setting "Allow switching to third party applications"](https://safeexambrowser.org/macosx/mac_usermanual_en.html#PermittedProcessesSection) isn't enabled (which might be only possible if OpenOlat allows to upload your own custom settings saved in SEB).

**Comment by danschlet** [comment]
You can use the email address info at safeexambrowser dot org to send confidential settings/access data.

**Comment by kollbrt** [comment]
I'll send you the SEB settings by Email.
The strange thing is, that with windows devices there is no problem with the same setting and the same SEB settings (we tested that several times).

**Comment by danschlet** [comment]
I couldn't yet do full-scale debugging, but I had a closer look at the settings after I saw some strange information in the log file. I don't think that these URL filter entries will work:

windows.open
jquery
js.plugin.min.js

**The expression in a URL filter always has to be a part of a URL**. SEB only filters URLs, not HTML elements. See the manual ([Filter section](https://safeexambrowser.org/macosx/mac_usermanual_en.html#NetworkPaneFilterSection)):

Expression: A Text field which contains the filtering expression or pattern, either in a regular expression (Regex) format or a simpler filter expression containing the wildcard char '*'. A filter expression can filter against all elements of a URL/URI according to [RFC 3986](http://tools.ietf.org/html/rfc3986): 

`scheme://user:password@host:port/path?query#fragment`

So if these filter expressions will be interpreted as URLs, they don't make sense.

**It's irrelevant that the same settings work in SEB for Windows.** SEB for Windows and SEB for macOS/iOS use different browser engines and different operating system APIs, which will behave differently in edge cases. **Also web applications use browser-specific code sometimes** (they detect which browser/browser engine is used and use different HTML, CSS or Javascript code in cases where that's technically necessary).

Relevant is only what the SEB manual (and specifications) state, those should be followed and it's not a good idea to experiment with not officially supported setting values. 


It would be great if you could retest your URL filters and most importantly:

- **If settings with URL filter don't work, test the same exam setup without URL filters** (URL filtering disabled). 
- **If it works without URL filtering, then you have to change your URL filters**. 
- If you feel that you cannot figure out working URL filters, **try using the interactive Teach SEB allowed/blocked URLs feature** (currently only available in SEB for macOS, see manual [Filter section](https://safeexambrowser.org/macosx/mac_usermanual_en.html#NetworkPaneFilterSection)).
- If it still doesn't work, then create a test case for us with SEB settings **without activated Kiosk Mode** (Windows: setting **Kiosk Mode: None**, see https://safeexambrowser.org/windows/win_usermanual_en.html#SecurityPane, macOS: setting **Allow switching to third party applications** enabled, see https://safeexambrowser.org/macosx/mac_usermanual_en.html#PermittedProcessesSection).

**URL filtering is a difficult to use feature**. Depending on the web site or web application used, it might not be possible to create URL filters which will work correctly. Maybe we could support more options in SEB for filtering web content (similar to what add blockers do), but given our backlog of feature requests I don't think this is realistic for the near future. For now we cannot give extensive support for creating URL filters, this is more of a DIY-feature.

Much better would be if vendors of web applications like web office apps would integrate a "locked exam mode" or something similar, which could be configured in their applications. For that I would file feature requests to these vendors.

In any case thanks for your feedback. If there is a bug in SEB, we're of course very thankful if community feedback helps to locate, analyze and fix it.
